README.pod
=pod

=encoding utf-8

=head1 NAME

Arriba - PSGI web server with SPDY support

=head1 VERSION

version 0.020

=head1 SYNOPSIS

Launch a plain HTTP server listening on port 5080:

    arriba --listen :5080

Launch an HTTPS server on port 5443, no SPDY:

    arriba --listen :5443:ssl --ssl-cert cert.pem --ssl-key key.pem

Launch an HTTPS server with SPDY support:

    arriba --listen :5443:ssl --ssl-cert cert.pem --ssl-key key.pem \
        --enable-spdy

=head1 DESCRIPTION

Arriba is a PSGI web server based on L<Starman> and sharing most of its
features, with added support for the SPDY protocol.

B<WARNING:> Arriba is still in early stage of development and is not ready for
production use.

=head1 SEE ALSO

Please see those modules/websites for more information related to this module.

=over 4

=item *

L<Starman>

=item *

L<Net::Server::PreFork>

=back

=head1 AUTHOR

Michal Wojciechowski <odyniec@cpan.org>

=head1 ACKNOWLEDGEMENTS

Basic server code and plain HTTP connection support is based on L<Starman>,
written by Tatsuhiko Miyagawa.

SPDY support is provided by L<Net::SPDY>, written by Lubomir Rintel.

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Michal Wojciechowski.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Arriba

{{$NEXT}}

0.020     2013-11-17
          First version, released on an unsuspecting world.


lib/Arriba/Server.pm
package Arriba::Server;

use warnings;
use strict;

use parent 'Net::Server::PreFork';

use HTTP::Date;
use HTTP::Status qw(status_message);
use HTTP::Parser::XS qw(parse_http_request);
use IO::Socket::SSL;

use Plack::Util;
use Plack::TempBuffer;

use constant DEBUG => $ENV{ARRIBA_DEBUG};
use constant CHUNKSIZE => 64 * 1024;

my $null_io = do { open my $io, "<", \""; $io };

use Net::Server::SIG qw(register_sig);

# Override Net::Server's HUP handling - just restart all the workers and that's
# about it
sub sig_hup {
    my $self = shift;
    $self->hup_children;
}

sub run {
    my ($self, $app, $options) = @_;

    $self->{app} = $app;
    $self->{options} = $options;

    my %extra = ();
    if ($options->{pid}) {
        $extra{pid_file} = $options->{pid};
    }
    if ($options->{daemonize}) {
        $extra{setsid} = $extra{background} = 1;
    }
    if ($options->{ssl_cert}) {
        $extra{SSL_cert_file} = $options->{ssl_cert};
    }
    if ($options->{ssl_key}) {
        $extra{SSL_key_file} = $options->{ssl_key};
    }
    if (!exists $options->{keepalive}) {
        $options->{keepalive} = 1;
    }
    if (!exists $options->{keepalive_timeout}) {
        $options->{keepalive_timeout} = 1;
    }

    my @port;
    for my $listen (@{$options->{listen} || [ "$options->{host}:$options->{port}" ]}) {
        my %listen;
        if ($listen =~ /:/) {
            my($h, $p, $opt) = split /:/, $listen, 3;
            $listen{host} = $h if $h;
            $listen{port} = $p;
            $listen{proto} = 'ssl' if 'ssl' eq lc $opt;
        }
        else {
            %listen = (
                host  => 'localhost',
                port  => $listen,
                proto => 'unix',
            );
        }
        push @port, \%listen;
    }

    my $workers = $options->{workers} || 5;

    local @ARGV = ();

    $self->SUPER::run(
        port => \@port,
        host => '*',
        proto => $options->{ssl} ? 'ssl' : 'tcp',
        serialize => 'flock',
        log_level => DEBUG ? 4 : 2,
        ($options->{error_log} ? ( log_file => $options->{error_log} ) : () ),
        min_servers => $options->{min_servers} || $workers,
        min_spare_servers => $options->{min_spare_servers} || $workers - 1,
        max_spare_servers => $options->{max_spare_servers} || $workers - 1,
        max_servers => $options->{max_servers} || $workers,
        max_requests => $options->{max_requests} || 1000,
        user => $options->{user} || $>,
        group => $options->{group} || $),
        listen => $options->{backlog} || 1024,
        check_for_waiting => 1,
        no_client_stdout => 1,
        %extra
    );
}

sub configure_hook {
    my $self = shift;

    # FIXME: Is this (configure_hook) the best place for this?
    
    if ($self->{options}->{listen_ssl}) {
        $self->{server}->{ssl_args}->{SSL_key_file} =
            $self->{options}->{ssl_key_file};
        $self->{server}->{ssl_args}->{SSL_cert_file} =
            $self->{options}->{ssl_cert_file};
    }

    $self->SUPER::configure_hook(@_);
}

sub pre_bind {
    my $self = shift;

    $self->SUPER::pre_bind(@_);

    if ($self->{options}->{spdy}) {
        # Enable SPDY on SSL sockets
        for my $sock (@{$self->{server}->{sock}}) {
            if ($sock->NS_proto eq 'SSL') {
                $sock->SSL_npn_protocols(['spdy/3']);
            }
        }
    }
}

sub pre_loop_hook {
    my $self = shift;
 
    my $port = $self->{server}->{port}->[0];
    my $proto = $port->{proto} eq 'ssl'  ? 'https' :
                $port->{proto} eq 'unix' ? 'unix' : 'http';

    $self->{options}{server_ready}->({
        host => $port->{host},
        port => $port->{port},
        proto => $proto,
        server_software => 'Arriba',
    }) if $self->{options}{server_ready};
 
    register_sig(
        TTIN => sub { $self->{server}->{$_}++ for qw( min_servers max_servers ) },
        TTOU => sub { $self->{server}->{$_}-- for qw( min_servers max_servers ) },
        QUIT => sub { $self->server_close(1) },
    );
}

sub server_close {
    my($self, $quit) = @_;
 
    if ($quit) {
        $self->log(2, $self->log_time . " Received QUIT. Running a graceful shutdown\n");
        $self->{server}->{$_} = 0 for qw( min_servers max_servers );
        $self->hup_children;
        while (1) {
            Net::Server::SIG::check_sigs();
            $self->coordinate_children;
            last if !keys %{$self->{server}{children}};
            sleep 1;
        }
        $self->log(2, $self->log_time . " Worker processes cleaned up\n");
    }
 
    $self->SUPER::server_close();
}

sub run_parent {
    my $self = shift;
    $0 = "arriba master " . join(" ", @{$self->{options}{argv} || []});
    no warnings 'redefine';
    local *Net::Server::PreFork::register_sig = sub {
        my %args = @_;
        delete $args{QUIT};
        Net::Server::SIG::register_sig(%args);
    };
    $self->SUPER::run_parent(@_);
}

sub child_init_hook {
    my $self = shift;
    srand();
    if ($self->{options}->{psgi_app_builder}) {
        $self->{app} = $self->{options}->{psgi_app_builder}->();
    }
    $0 = "arriba worker " . join(" ", @{$self->{options}{argv} || []});
}
 
sub post_accept_hook {
    my $self = shift;
 
    $self->{client} = { };
}

sub process_request {
    my $self = shift;

    my $client = $self->{server}->{client};

    # Is this an SSL connection?
    my $ssl = $client->NS_proto eq 'SSL';

    if ($ssl && $client->next_proto_negotiated &&
        $client->next_proto_negotiated eq 'spdy/3')
    {
        # SPDY connection
        require Arriba::Connection::SPDY;
        $self->{client}->{connection} =
            Arriba::Connection::SPDY->new($client);
    }
    else {
        # HTTP(S) connection
        require Arriba::Connection::HTTP;
        $self->{client}->{connection} =
            Arriba::Connection::HTTP->new($client, ssl => $ssl,
                chunk_size => CHUNKSIZE);
    }

    my $connection = $self->{client}->{connection};

    while (my $req = $connection->read_request) {
        my $env;
        my $conn_header;

        if ($req->{env}) {
            # Headers already parsed
            $env = $req->{env};
        }
        else {
            $env = {
                REMOTE_ADDR => $self->{server}->{peeraddr},
                REMOTE_HOST => $self->{server}->{peerhost} || $self->{server}->{peeraddr},
                REMOTE_PORT => $self->{server}->{peerport} || 0,
                SERVER_NAME => $self->{server}->{sockaddr} || 0, # XXX: needs to be resolved?
                SERVER_PORT => $self->{server}->{sockport} || 0,
                SCRIPT_NAME => '',
                'psgi.version' => [ 1, 1 ],
                'psgi.errors' => *STDERR,
                'psgi.url_scheme' => $req->{scheme},
                'psgi.nonblocking' => Plack::Util::FALSE,
                'psgi.streaming' => Plack::Util::TRUE,
                'psgi.run_once' => Plack::Util::FALSE,
                'psgi.multithread' => Plack::Util::FALSE,
                'psgi.multiprocess' => Plack::Util::TRUE,
                'psgix.io' => $client,
                'psgix.input.buffered' => Plack::Util::TRUE,
                'psgix.harakiri' => Plack::Util::TRUE,
            };

            my $reqlen = parse_http_request($req->{headers}, $env);

            if ($reqlen < 0) {
                # Bad request
                $self->_http_error($req, 400);
                last;
            }

            $conn_header = delete $env->{HTTP_CONNECTION};
            my $proto = $env->{SERVER_PROTOCOL};

            if ($proto && $proto eq 'HTTP/1.0' ) {
                if ($conn_header && $conn_header =~ /^keep-alive$/i) {
                    # Keep-alive only with explicit header in HTTP/1.0
                    $connection->{_keepalive} = 1;
                }
                else {
                    $connection->{_keepalive} = 0;
                }
            }
            elsif ($proto && $proto eq 'HTTP/1.1') {
                if ($conn_header && $conn_header =~ /^close$/i ) {
                    $connection->{_keepalive} = 0;
                }
                else {
                    # Keep-alive assumed in HTTP/1.1
                    $connection->{_keepalive} = 1;
                }
 
                # Do we need to send 100 Continue?
                if ($env->{HTTP_EXPECT}) {
                    if ($env->{HTTP_EXPECT} eq '100-continue') {
                        # FIXME:
                        #syswrite $client, 'HTTP/1.1 100 Continue' . $CRLF . $CRLF;
                    }
                    else {
                        $self->_http_error(417, $env);
                        last;
                    }
                }
 
                unless ($env->{HTTP_HOST}) {
                    # No host, bad request
                    $self->_http_error(400, $env);
                    last;
                }
            }

            unless ($self->{options}->{keepalive}) {
                $connection->{_keepalive} = 0;
            }

            $req->{env} = $env;
        }
        
        # Process this request later if it's not ready yet
        next if !$req->{complete};

        if ($req->{body_stream}) {
            $env->{'psgi.input'} = $req->{body_stream}->rewind;
        }
        else {
            $env->{'psgi.input'} = $null_io;
        }

        my $res = Plack::Util::run_app($self->{app}, $env);

        if (ref $res eq 'CODE') {
            $res->(sub { $connection->write_response($req, $_[0]) });
        }
        else {
            $connection->write_response($req, $res);
        }

        my $sel = IO::Select->new($client);
        last unless $sel->can_read($self->{options}->{keepalive_timeout});
    }
}

sub _http_error {
    my ($self, $req, $code) = @_;
 
    my $status = $code || 500;
    my $msg = status_message($status);
 
    my $res = [
        $status,
        [ 'Content-Type' => 'text/plain', 'Content-Length' => length($msg) ],
        [ $msg ],
    ];

    $self->{client}->{connection}->{_keepalive} = 0;
    $self->{client}->{connection}->write_response($req, $res);
}

1;



lib/Arriba/Connection/HTTP.pm
package Arriba::Connection::HTTP;

use warnings;
use strict;

use Data::Dump qw(dump);
use HTTP::Status qw(status_message);
use IO::Socket qw(:crlf);
use Plack::Util;
use Socket qw(IPPROTO_TCP TCP_NODELAY);

use parent 'Arriba::Connection';

sub new {
    my $class = shift;
    my $self = $class->SUPER::new(@_);

    if ($self->{client}->NS_proto eq 'TCP') {
        setsockopt($self->{client}, IPPROTO_TCP, TCP_NODELAY, 1)
            or die $!;
    }
    
    $self->{_inputbuf} = '';
    $self->{_current_req} = undef;
    $self->{_keepalive} = 1;

    return $self;
}

sub read_request {
    my $self = shift;

    my $req;

    if ($req = $self->{_current_req}) {
        # Partially processed request
        my $get_chunk = sub {
            if ($self->{_inputbuf}) {
                my $chunk = delete $self->{_inputbuf};
                return ($chunk, length $chunk);
            }
            my $read = sysread $self->{client}, my($chunk), $self->{chunk_size};
            return ($chunk, $read);
        };

        my $chunked = do {
            no warnings;
            lc delete $req->{env}->{HTTP_TRANSFER_ENCODING} eq 'chunked'
        };

        if ((my $cl = $req->{content_length}) >= 0) {
            $req->{body_stream} = Stream::Buffered->new($req->{content_length});
            while ($cl > 0) {
                my($chunk, $read) = $get_chunk->();

                if (!defined $read || $read == 0) {
                    die "Read error: $!\n";
                }

                $cl -= $read;
                $req->{body_stream}->print($chunk);
            }
        }
        elsif ($chunked) {
            $req->{body_stream} = Stream::Buffered->new;
            my $chunk_buffer = '';
            my $length;

            DECHUNK:
            while (1) {
                my($chunk, $read) = $get_chunk->();
                $chunk_buffer .= $chunk;

                while ($chunk_buffer =~ s/^(([0-9a-fA-F]+).*\015\012)// ) {
                    my $trailer = $1;
                    my $chunk_len = hex $2;
 
                    if ($chunk_len == 0) {
                        last DECHUNK;
                    } elsif (length $chunk_buffer < $chunk_len + 2) {
                        $chunk_buffer = $trailer . $chunk_buffer;
                        last;
                    }
 
                    $req->{body_stream}->print(substr($chunk_buffer, 0,
                        $chunk_len, ''));
                    $chunk_buffer =~ s/^\015\012//;
 
                    $length += $chunk_len;
                }
 
                last unless $read && $read > 0;
            }
 
            $req->{content_length} = $length;
        }

        $req->{complete} = 1;
        $self->{_current_req} = undef;
    }
    elsif ($self->{_keepalive}) {
        # New request
        $req = Arriba::Request->new($self);
        $req->{scheme} = $self->{ssl} ? 'https' : 'http';

        while (1) {
            last if defined $self->{_inputbuf} &&
                $self->{_inputbuf} =~ /$CRLF$CRLF/s;
            
            my $read = sysread $self->{client}, my $buf, $self->{chunk_size};

            if (!defined $read || $read == 0) {
                die "Read error: $!\n";
            }

            $self->{_inputbuf} .= $buf;
        }

        (my $headers, $self->{_inputbuf}) =
            split /$CRLF$CRLF/, $self->{_inputbuf}, 2;

        # Add back two CRLFs, HTTP::Parser's parse_http_requests expects that
        $req->{headers} = $headers . $CRLF . $CRLF;

        if ($req->{headers} =~ /^content-length:\s*(\d+)\015?$/im) {
            $req->{content_length} = $1;
            $self->{_current_req} = $req;
        }
        else {
            # No "Content-length" header, we have the whole request
            $req->{complete} = 1;
            $self->{_current_req} = undef;
        }
    }

    return $req;
}

sub write_response {
    my $self = shift;
    my $req = shift;
    my $res = shift;

    my $proto = $req->{env}->{SERVER_PROTOCOL};
    my $status = $res->[0];

    my %headers;
    my $chunked;
    my @header_lines = ("$proto $status " . status_message($status));
    
    my $res_headers = $res->[1];

    for (my $i = 0; $i < @$res_headers; $i += 2) {
        next if $res_headers->[$i] eq 'Connection';
        push @header_lines, $res_headers->[$i] . ": " . $res_headers->[$i+1];
        $headers{lc $res_headers->[$i]} = $res_headers->[$i+1];
    }

    if ($proto eq 'HTTP/1.1') {
        if (!exists $headers{'content-length'}) {
            if ($status !~ /^1\d\d|[23]04$/) {
                push @header_lines, 'Transfer-Encoding: chunked';
                $chunked = 1;
            }
        }
        elsif (my $te = $headers{'transfer-encoding'}) {
            if ($te eq 'chunked') {
                $chunked = 1;
            }
        }
    }
    else {
        if (!exists $headers{'transfer-encoding'}) {
            $self->{_keepalive} = 0;
        }
    }

    if ($self->{_keepalive}) {
        push @header_lines, 'Connection: keep-alive';
    }
    else {
        push @header_lines, 'Connection: close';
    }

    syswrite $self->{client}, join($CRLF, @header_lines, '') . $CRLF;

    if (defined $res->[2]) {
        Plack::Util::foreach($res->[2], sub {
            my $buffer = $_[0];
            my ($len, $offset);

            if ($chunked) {
                $len = length $buffer;
                return unless $len;
                $buffer = sprintf( "%x", $len ) . $CRLF . $buffer . $CRLF;
            }
            
            $len = length $buffer;
            $offset = 0;
            while ($len) {
                my $written = syswrite $self->{client}, $buffer, $len, $offset;
                # TODO: Handle errors maybe?
                $len -= $written;
                $offset += $written;
            }
        });
 
        syswrite $self->{client}, "0$CRLF$CRLF" if $chunked;
    }
    else {
        # TODO: Above loop also needed here
        return Plack::Util::inline_object
            write => sub {
                my $buf = $_[0];
                if ($chunked) {
                    my $len = length $buf;
                    return unless $len;
                    $buf = sprintf( "%x", $len ) . $CRLF . $buf . $CRLF;
                }
                syswrite $self->{client}, $buf;
            },
            close => sub {
                syswrite $self->{client}, "0$CRLF$CRLF" if $chunked;
            };
    }
}

1;



lib/Arriba/Connection/SPDY.pm
package Arriba::Connection::SPDY;

use warnings;
use strict;

use HTTP::Date;
use HTTP::Status qw(status_message);
use IO::Socket qw(:crlf);
use Net::SPDY::Session;
use Plack::Util;

use Arriba::Request;

use parent 'Arriba::Connection';

# The current version of Net::SPDY::Session has some debugging code in the
# process_frame method, which breaks things. Until that gets fixed, we'll use
# our own version of this method.
{
    ## no critic
    no strict 'refs';
    no warnings 'redefine';
    *{"Net::SPDY::Session::process_frame"} = sub {
        my $self = shift;

        my %frame = $self->{framer}->read_frame ();
        return () unless %frame;

        if (not $frame{control}) {
            #warn 'Not implemented: Data frame received';
            return %frame;
        }

        if ($frame{type} == Net::SPDY::Framer::SYN_STREAM) {
        } elsif ($frame{type} == Net::SPDY::Framer::SETTINGS) {
            $self->got_settings (%frame);
        } elsif ($frame{type} == Net::SPDY::Framer::PING) {
            $self->{framer}->write_ping (
                flags => 0,
                id => $frame{id},
            );
        } elsif ($frame{type} == Net::SPDY::Framer::GOAWAY) {
            $self->close (0);
        } elsif ($frame{type} == Net::SPDY::Framer::HEADERS) {
            # We should remember values gotten here for stream
            #warn 'Not implemented: Got headers frame'
        } elsif ($frame{type} == Net::SPDY::Framer::WINDOW_UPDATE) {
        } else {
            die 'Unknown frame type '.$frame{type};
        }

        return %frame;
    };
}

sub new {
    my $class = shift;
    my $self = $class->SUPER::new(@_);

    $self->{_spdy} = {
        session => Net::SPDY::Session->new($self->{client}),
        streams => {},
    };

    return $self;
}

sub read_request {
    my $self = shift;

    my $session = $self->{_spdy}->{session};
    my $streams = $self->{_spdy}->{streams};

    my $req;

    while (my %frame = $session->process_frame) {
        if (exists $frame{type}) {
            if ($frame{type} == Net::SPDY::Framer::SYN_STREAM) {
                # Request initiated
                $req = Arriba::Request->new($self);
                $req->{_spdy} = {
                    # Keep track of this request's stream ID
                    stream_id => $frame{stream_id},
                    # Save a reference to the frame that initiated this request
                    frame => \%frame,
                };

                $streams->{$frame{stream_id}} = { req => $req };

                my %frame_headers = @{$frame{headers}};
                my @http_headers = @{$frame{headers}};

                my $headers = '';
                # Construct the HTTP request line
                $headers .= $frame_headers{':method'}  . ' ' .
                    $frame_headers{':path'} . ' ' .
                    $frame_headers{':version'} . $CRLF;

                $req->{scheme} = $frame_headers{':scheme'};

                if ($frame_headers{':host'}) {
                    $headers .= 'host: ' . $frame_headers{':host'} . $CRLF;
                }
                
                for (my $i = 0; $i < $#http_headers; $i += 2) {
                    if ($http_headers[$i] !~ /^:/) {
                        $headers .= "$http_headers[$i]: $http_headers[$i+1]$CRLF";
                    }
                }

                $headers .= $CRLF;

                $req->{headers} = $headers;
            }
            else {
                # FIXME: Other cases
                next;
            }
        }
        # Data frame - check for existing stream ID
        elsif (my $stream = $streams->{$frame{stream_id}}) {
            # Grab the corresponding request
            $req = $stream->{req};
        }
        # Unknown stream ID
        else {
            # TODO: Error
            next;
        }

        if (!$frame{control}) {
            # TODO: Move the above check for existing stream ID in here?
            # Data frame
            if (!$req->{body_stream}) {
                # TODO: Initialize with content length?
                $req->{body_stream} = Stream::Buffered->new;
            }
            $req->{body_stream}->print($frame{data});
        }

        if ($frame{flags} & Net::SPDY::Framer::FLAG_FIN) {
            # Last frame on this stream
            $req->{complete} = 1;
        }

        return $req;
    }
}

sub write_response {
    my $self = shift;
    my $req = shift;
    my $res = shift;

    my $status = $res->[0];

    my %frame_headers = (
        ':version' => 'HTTP/1.1',
        ':status' => "$status " . status_message($status),
        'date' => HTTP::Date::time2str(),
        'server' => 'arriba', # FIXME
    );

    Plack::Util::header_iter($res->[1], sub {
        my ($k, $v) = @_;
        if (exists $frame_headers{lc $k}) {
            if (ref $frame_headers{lc $k} ne 'ARRAY') {
                $frame_headers{lc $k} = [ $frame_headers{lc $k} ];
            }
            push @{$frame_headers{lc $k}}, $v;
        }
        else {
            $frame_headers{lc $k} = $v;
        }
    });

    my %frame = (
        type => Net::SPDY::Framer::SYN_REPLY,
        stream_id => $req->{_spdy}->{stream_id},
        headers => [ %frame_headers ],
        control => 1,
        flags => 0,
    );

    my $framer = $self->{_spdy}->{session}->{framer};

    if (defined $res->[2]) {
        Plack::Util::foreach($res->[2], sub {
            # Send the previous frame
            $framer->write_frame(%frame);

            %frame = (
                stream_id => $req->{_spdy}->{stream_id},
                data => $_[0],
                control => 0,
                flags => 0,
            );
        });

        $frame{flags} |= Net::SPDY::Framer::FLAG_FIN;
        $framer->write_frame(%frame);
    }
    else {
        return Plack::Util::inline_object
            write => sub {
                # Send the previous frame
                $framer->write_frame(%frame);

                %frame = (
                    stream_id => $req->{_spdy}->{stream_id},
                    data => $_[0],
                    control => 0,
                    flags => 0,
                );
            },
            close => sub {
                $frame{flags} |= Net::SPDY::Framer::FLAG_FIN;
                $framer->write_frame(%frame);
            };
    }
}

1;



lib/Arriba.pm
package Arriba;

use strict;

# ABSTRACT: PSGI web server with SPDY support

# VERSION

1;
__END__

=head1 SYNOPSIS

Launch a plain HTTP server listening on port 5080:

    arriba --listen :5080

Launch an HTTPS server on port 5443, no SPDY:

    arriba --listen :5443:ssl --ssl-cert cert.pem --ssl-key key.pem

Launch an HTTPS server with SPDY support:

    arriba --listen :5443:ssl --ssl-cert cert.pem --ssl-key key.pem \
        --enable-spdy

=head1 DESCRIPTION

Arriba is a PSGI web server based on L<Starman> and sharing most of its
features, with added support for the SPDY protocol.

B<WARNING:> Arriba is still in early stage of development and is not ready for
production use.

=head1 ACKNOWLEDGEMENTS

Basic server code and plain HTTP connection support is based on L<Starman>,
written by Tatsuhiko Miyagawa.

SPDY support is provided by L<Net::SPDY>, written by Lubomir Rintel.

=head1 SEE ALSO

L<Starman>
L<Net::Server::PreFork>

=cut
